subprojects/core/src/main/java/org/gradle/execution/ProjectExecutionServices.java

```diff
@@ -115,24 +120,31 @@ TaskExecuter createTaskExecuter(TaskArtifactStateRepository repository,
 
         boolean buildCacheEnabled = buildCacheController.isEnabled();
         boolean scanPluginApplied = buildScanPlugin.isBuildScanPluginApplied();
+        TaskCacheKeyCalculator cacheKeyCalculator = new DefaultTaskCacheKeyCalculator();
 
         TaskExecuter executer = new ExecuteActionsTaskExecuter(
             buildCacheEnabled,
+            fingerprinterRegistry,
+            executionHistoryStore,
+            outputFilesRepository,
             buildOperationExecutor,
             asyncWorkTracker,
             actionListener,
             workExecutor
         );
+        executer = new ResolveIncrementalChangesTaskExecuter(executer);
         executer = new ResolveTaskOutputCachingStateExecuter(buildCacheEnabled, relativeFilePathResolver, executer);
         if (buildCacheEnabled || scanPluginApplied) {
-            executer = new ResolveBuildCacheKeyExecuter(executer, buildOperationExecutor, buildCacheController.isEmitDebugLogging());
+            executer = new ResolveBuildCacheKeyExecuter(buildOperationExecutor, cacheKeyCalculator, buildCacheController.isEmitDebugLogging(), executer);
         }
+        executer = new ResolveBeforeExecutionStateTaskExecuter(classLoaderHierarchyHasher, valueSnapshotter, fingerprinterRegistry, executer);
         executer = new ValidatingTaskExecuter(executer);
         executer = new SkipEmptySourceFilesTaskExecuter(inputsListener, executionHistoryStore, cleanupRegistry, outputChangeListener, executer, buildInvocationScopeId);
-        executer = new ResolvePreviousStateExecuter(executionHistoryStore, executer);
+        executer = new ResolveBeforeExecutionOutputsTaskExecuter(fingerprinterRegistry, executer);
```

Comment:
I guess we only need the output snapshots before the `SkipEmptySourceFilesTaskExecuter` if we have a task history. If we don't, there is no need to detect overlapping outputs for cleanup there.

Reply:
What are you suggesting to change here?

Reply:
That we could only resolve the before execution outputs when required in `SkipEmptySourceFilesTaskExecuter`. That would require some laziness for resolving the execution outputs again - so I guess we shouldn't do it.

The change to not resolve the outputs when there are no source files and no previous task history could speed up that step.

Reply:
How would that be useful? I mean, the only reason why we _wouldn't_ observe the outputs in `SkipEmptySourceFilesTaskExecuter` is if we continue executing the task further down the line. On that path the only case where we can finish without needing the output snapshots is if `ValidatingTaskExecuter` fails. That's a case I don't think is worth optimizing for.

Reply:
We also don't need to observe the outputs if there is no task history and no sources. So the optimization would be for the no-source case.

Reply:
What we could do is split the empty case check into two, and do this special case before. But I don't think it's worth it at this point. If you do, please create an issue so we can discuss in detail.